src/Processors/LimitTransform.cpp

```diff
@@ -226,27 +229,106 @@ LimitTransform::Status LimitTransform::preparePair(PortsData & data)
         input.setNeeded();
         return Status::NeedData;
     }
-
-    if (rows <= std::numeric_limits<UInt64>::max() - offset && rows_read >= offset + rows
-        && !limit_is_unreachable && rows_read <= offset + limit)
+    if (is_limit_positive)
     {
-        /// Return the whole chunk.
+        if (rows <= std::numeric_limits<UInt64>::max() - offset && rows_read >= offset + rows
+            && !limit_is_unreachable && rows_read <= offset + limit)
+        {
+            /// Return the whole chunk.
 
-        /// Save the last row of current chunk to check if next block begins with the same row (for WITH TIES).
-        if (with_ties && rows_read == offset + limit)
-            previous_row_chunk = makeChunkWithPreviousRow(data.current_chunk, data.current_chunk.getNumRows() - 1);
+            /// Save the last row of current chunk to check if next block begins with the same row (for WITH TIES).
+            if (with_ties && rows_read == offset + limit)
+                previous_row_chunk = makeChunkWithPreviousRow(data.current_chunk, data.current_chunk.getNumRows() - 1);
+        }
+        else
+            /// This function may be heavy to execute in prepare. But it happens no more than twice, and make code simpler.
+            splitChunk(data);
     }
     else
-        /// This function may be heavy to execute in prepare. But it happens no more than twice, and make code simpler.
-        splitChunk(data);
+    {
+        if (rows == 8192 || rows == 16384 || rows == 32768 || rows == 65536 || rows == 65505)
```

Comment:
I want to know if there is a way to locate the last block. Now I can only use this clumsy method to identify, which is not accurate.

Reply:
@Avogar @alexey-milovidov